core/src/HPX/Kokkos_HPX_WorkGraphPolicy.hpp

```diff
@@ -35,53 +35,28 @@ class ParallelFor<FunctorType, Kokkos::WorkGraphPolicy<Traits...>,
   Policy m_policy;
   FunctorType m_functor;
 
-  template <class TagType>
-  std::enable_if_t<std::is_void<TagType>::value> execute_functor(
-      const std::int32_t w) const noexcept {
-    m_functor(w);
-  }
-
-  template <class TagType>
-  std::enable_if_t<!std::is_void<TagType>::value> execute_functor(
-      const std::int32_t w) const noexcept {
-    const TagType t{};
-    m_functor(t, w);
-  }
-
  public:
-  void execute() const {
-    dispatch_execute_task(this, m_policy.space());
-    m_policy.space().fence(
-        "Kokkos::Experimental::Impl::HPX::ParallelFor<WorkGraphPolicy>: fence "
-        "after kernel execution");
+  void execute_range(int) const {
+    std::int32_t w = m_policy.pop_work();
+    while (w != Policy::COMPLETED_TOKEN) {
+      if (w != Policy::END_TOKEN) {
+        if constexpr (std::is_same_v<WorkTag, void>) {
+          m_functor(w);
+        } else {
+          m_functor(WorkTag{}, w);
+        }
+        m_policy.completed_work(w);
+      }
+
+      w = m_policy.pop_work();
+    }
   }
 
-  void execute_task() const {
-    // See [note 1] in Kokkos_HPX.hpp for an explanation. The work graph policy
-    // does not store an execution space instance, so we only need to reset the
-    // parallel region count here.
-    Kokkos::Experimental::HPX::reset_count_on_exit_parallel reset_count_on_exit;
-
+  void execute() const {
     const int num_worker_threads = Kokkos::Experimental::HPX::concurrency();
```

Comment:
https://github.com/kokkos/kokkos/issues/5655
```suggestion
    const int num_worker_threads = m_policy.space().concurrency();
```

Reply:
That said, this can be fixed as a follow-up.